lib/dma.h

```diff
@@ -182,24 +211,33 @@ _dma_mark_dirty(const dma_controller_t *dma, const dma_memory_region_t *region,
     range_to_pages(sg->offset, sg->length, dma->dirty_pgsize,
                    &pgstart, &pgend);
 
-    index = bit_to_u8(pgstart);
-    end = bit_to_u8(pgend) + !!(bit_to_u8off(pgend));
-
-    for (i = index; i < end; i++) {
-        uint8_t bm = ~0;
-
-        /* Mask off any pages in the first u8 that aren't in the range. */
-        if (i == index && bit_to_u8off(pgstart) != 0) {
-            bm &= ~((1 << bit_to_u8off(pgstart)) - 1);
-        }
+    bm_sg_start = bit_to_u8(pgstart);
+    bm_sg_end = bit_to_u8(pgend) + !!(bit_to_u8off(pgend));
+
+    assert(bm_sg_start <= bm_sg_end);
+    
+    u8_set_all(&bm);
+    /* If only one byte of the bitmap has dirty pages, mark them and return. */
+    if (pgstart == pgend || bm_sg_start == bm_sg_end - 1) {
+        mask_any_lower_bits(&bm, pgstart);
+        mask_any_higher_bits(&bm, pgend);
+        apply_byte_mask(&region->dirty_bitmap[bm_sg_start], bm);
+        return;
+    }
 
-        /* Mask off any pages in the last u8 that aren't in the range. */
-        if (i == end - 1 && bit_to_u8off(pgend) != 0) {
-            bm &= ((1 << bit_to_u8off(pgend)) - 1);
-        }
+    /* Mask off any pages in the first u8 that aren't in the range. */
+    mask_any_lower_bits(&bm, pgstart);
+    apply_byte_mask(&region->dirty_bitmap[bm_sg_start], bm);
 
-        __atomic_or_fetch(&region->dirty_bitmap[i], bm, __ATOMIC_SEQ_CST);
+    /* Set 1s for all of the pages in the middle u8s of the range. */
+    u8_set_all(&bm);
+    for (i = bm_sg_start + 1; i < bm_sg_end - 1; i++) {
+        apply_byte_mask(&region->dirty_bitmap[i], bm);
```

Comment:
No need for a loop, this can be a memset() now!

Reply:
AH yes, I think it can!